blog/_posts/2021-02-16-improving-scala-library-ecosystem.md

```diff
@@ -0,0 +1,127 @@
+---
+layout: blog-detail
+post-type: blog
+by: Eugene Yokota, Julien Richard-Foy
+title: Improving the Scala Library Ecosystem
+---
+
+One of the things that makes Scala powerful and fun to use is its library ecosystem — spanning across the Maven ecosystem for JVM, and npm for Scala.JS. We can use libraries to do operations like download files from the web, or we can use it to adopt different programming paradigms. The library ecosystem allows us to write more useful programs with less effort.
+
+## What is a version conflict?
+
+However, the library ecosystem is not without problems. A library that you pulled could depend on other libraries, and the transitive dependencies could cause version conflicts. Here's a quick example of a Scala project that uses Akka HTTP, a Postgres database, and JSON. Its build declares two library dependencies, `akka-http-circe` and `doobie-postgres-circe`:
+
+~~~ scala
+libraryDependencies ++= Seq(
+  "de.heikoseeberger" %% "akka-http-circe"       % "1.26.0",
+  "org.tpolecat"      %% "doobie-postgres-circe" % "0.10.0"
+)
+~~~
+
+Note that both libraries use Circe for working with JSON.
+
+sbt loads the project build, the project compiles successfully, but when you try to run it, you may get a `NoSuchMethodError` exception at run-time!
+
+What happens here is that one of the transitive dependencies of `akka-http-circe` depends on `circe-core` version 0.11.1, and another transitive dependency of `doobie-postgres-circe` depends on `circe-core` version 0.13.0. Here's an excerpt of the output of sbt’s `dependencyTree` task:
+
+~~~ text
+[info]   +-de.heikoseeberger:akka-http-circe_2.12:1.26.0 [S]
+[info]   | +-io.circe:circe-core_2.12:0.11.1 (evicted by: 0.13.0)
+[info]   |     
+[info]   +-org.tpolecat:doobie-postgres-circe_2.12:0.10.0 [S]
+[info]     +-io.circe:circe-core_2.12:0.13.0 [S]
+~~~
+
+Unfortunately, these two versions of Circe are not binary compatible. Such version conflicts are inevitable in a dependency graph of any practical size. The problem is that we carefully program our code using a statically checked type system, but when it comes to production code we accept swapping out the JAR file with something that our dependency resolver like Coursier and Apache Ivy selected on a whim.
```

Comment:
Nitpick: "inevitable" seems exaggerated to me. Most of the time you get lucky, though often enough you don't! How about "common"?

Reply:
Given some decent sized graph, having a version conflict is inevitable. You're totally right about most of the time picking the latest version does the ok thing, and it's only _sometimes_ we would hit binary incompatible behavior such as `NoSuchMethodError` during testing or production.